src/simulation/point.py

```diff
@@ -178,17 +176,26 @@ def process_advection(self, delta_time: float) -> None:
                 crossing_point = (round(x + i * advection_x / abs(advection_y)), y + i * advection_y / abs(advection_y))
             if self.engine.get_topography(crossing_point) == TopographyState.LAND:
                 next_x, next_y = map(int, crossing_point)
-                # to ensure that buffer will be zeroed in a next step, because oil can't go further
-                self.advection_buffer = np.array([next_x - x, next_y - y], dtype='float64')
+                advection_vector = [0, 0]  # so it's not going through land in the next step
                 break
 
-        if 0 <= next_x < const.point_side_count and 0 <= next_y < const.point_side_count:
-            if (next_x, next_y) not in self.world:
-                self.world[(next_x, next_y)] = Point((next_x, next_y), self.initial_values, self.engine)
-                self.engine.points_changed.append((next_x, next_y))
-            self.world[(next_x, next_y)].oil_buffer.append((self.oil_mass, self.viscosity_dynamic, self.emulsification_rate))
-        self.advection_buffer -= np.array([next_x - x, next_y - y])
-        self.oil_mass = 0
+        fractional_part_x = advection_vector[0] % 1
+        fractional_part_y = advection_vector[1] % 1
+        x_shift = int(fractional_part_x / abs(fractional_part_x)) if fractional_part_x != 0 else 0
+        y_shift = int(fractional_part_y / abs(fractional_part_y)) if fractional_part_y != 0 else 0
+        neighbours = [(next_x, next_y + y_shift), (next_x + x_shift, next_y), (next_x + x_shift, next_y + y_shift)]
+        areas = [fractional_part_y * (1 - fractional_part_x), fractional_part_x * (1 - fractional_part_y),
+                 fractional_part_x * fractional_part_y]
+        if self.initial_values.neighbourhood == Neighbourhood.VON_NEUMANN:
+            neighbours.pop()
+            area_to_split = areas.pop()
+            areas[0] += area_to_split * areas[0] / (areas[0] + areas[1]) if areas[0] + areas[1] != 0 else 0
+            areas[1] += area_to_split * areas[1] / (areas[0] + areas[1]) if areas[0] + areas[1] != 0 else 0
+        for neighbour, area in zip(neighbours, areas):
+            self.move_oil_to_other(neighbour, self.oil_mass * area)
```

Comment:
we should remeber value of self.oil_mass before we do all of moveing -> because after first move self.oil_mass changes